src/lib.rs

```diff
@@ -0,0 +1,464 @@
+
+        if !failures.is_empty() {
+            process::exit(1);
+        }
+    }
+
+    fn parse_tests(&self, test_str: &str) -> HashMap<String, Test> {
+        let mut tests = HashMap::new();
+        let parse_toml = test_str
+            .trim()
+            .parse::<Value>()
+            .expect("Couldn't parse test TOML");
+        match parse_toml {
+            Value::Table(t) => {
+                for (cmd_name, test_entries) in t {
+                    match tests.entry(cmd_name.to_lowercase()) {
+                        Entry::Occupied(_) => {
+                            panic!("Test TOML key '{}' is specified more than once.", cmd_name)
+                        }
+                        Entry::Vacant(e) => match test_entries {
+                            Value::Table(t) => {
+                                let mut status = None;
+                                let mut stderr = None;
+                                let mut stdout = None;
+                                for (k, v) in t {
+                                    let v = v
+                                        .as_str()
+                                        .expect(&format!(
+                                            "Test TOML key '{}' has a non-string value.",
+                                            k
+                                        ))
+                                        .trim();
+                                    match k.as_str() {
+                                        "status" => {
+                                            let v_lower = v.to_lowercase();
+                                            if v_lower != "success" && v_lower != "error" {
+                                                panic!("Unknown status '{}'", v);
+                                            }
+                                            status = Some(v_lower);
+                                        }
+                                        "stderr" => {
+                                            stderr = Some(v.to_owned());
+                                        }
+                                        "stdout" => {
+                                            stdout = Some(v.to_owned());
+                                        }
+                                        _ => panic!("Test TOML key '{}' unknown.", v),
+                                    }
+                                }
+                                e.insert(Test {
+                                    status,
+                                    stderr,
+                                    stdout,
+                                });
+                            }
+                            _ => panic!("TOML entry {} must contain a table.", cmd_name),
+                        },
+                    }
+                }
+            }
+            _ => panic!("Test TOML must be a table."),
+        }
+
```

Comment:
Whitespace?

Reply:
That one was intentional, just to differentiate it from the massive `match` above. But I don't feel strongly about it. Since it looks like I'm going to rewrite the parser, I'm not sure it's too important.